### PR TITLE
8308286 Fix clang warnings in linux code

### PR DESCRIPTION
**Pull request by savoptik**
When using the clang compiler to build OpenJDk on Linux, we encounter various "warnings as errors".
They can be fixed with small changes.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8308286](https://bugs.openjdk.org/browse/JDK-8308286): Fix clang warnings in linux code (**Bug** - P4)

### Reviewers
 * [Alexey Ushakov](https://openjdk.org/census#avu) (@avu - Committer) ⚠️ Review applies to [b423bd4c](https://git.openjdk.org/jdk/pull/14033/files/b423bd4cdbefecec7468d8538819d39372613fca)
 * [Daniel Jeliński](https://openjdk.org/census#djelinski) (@djelinski - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14033/head:pull/14033` \
`$ git checkout pull/14033`

Update a local copy of the PR: \
`$ git checkout pull/14033` \
`$ git pull https://git.openjdk.org/jdk.git pull/14033/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14033`

View PR using the GUI difftool: \
`$ git pr show -t 14033`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14033.diff">https://git.openjdk.org/jdk/pull/14033.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14033#issuecomment-1551314451)

### CODE REVIEW

**Comment by avu** [comment]
I would suggest either disable warnings on per file basis or rewrite problematic code. I seems that disabling warnings per fragment of code makes it less readable.

**Review by prrace — CHANGES_REQUESTED** [review]
I don't like this approach at all. if github had a "reject" button I'd be pushing it now.
updating the makefiles is the normal way to do this but I don't know if we even want that.
Until clang is the supported compiler for Linux then I see no reason for this to be in JDK at all.

Code changes which make it so there's no warning would be the preferred way but I'd expect that to be tested properly. Also if the warning is spurious - sometimes they are - then that would NOT be appropriate.

**Comment by djelinski** [comment]
According to our docs, [clang is a supported compiler for Linux](https://github.com/openjdk/jdk/blob/master/doc/building.md#native-compiler-toolchain-requirements).

Here's an example how warning exclusion is implemented today:
https://github.com/openjdk/jdk/blob/master/make/modules/java.desktop/lib/Awt2dLibraries.gmk#L827

**Comment by kimbarrett** [comment]
> According to our docs, [clang is a supported compiler for Linux](https://github.com/openjdk/jdk/blob/master/doc/building.md#native-compiler-toolchain-requirements).

I think that's aspirational rather than actual.  Clang has been included in
that list for a long time (since at least JDK 9), but I think there's never
been anyone claiming to provide such support, be a maintainer, or even
build/test on some regular basis:
https://wiki.openjdk.org/display/Build/Supported+Build+Platforms

File: src/jdk.hotspot.agent/linux/native/libsaproc/ps_core.c
Comment (kimbarrett) [inline src/jdk.hotspot.agent/linux/native/libsaproc/ps_core.c:575]: What problem is being "fixed" by these?  I'm dubious that this is the best solution to whatever the problem
is, but can't evaluate or suggest alternatives without knowing what it is.

**Review by kimbarrett — CHANGES_REQUESTED** [review]
All of the -Wformat-nonliteral changes make me wonder why we're seeing these
with clang but not with gcc.  Figuring that out will likely give a better
chance of acceptance.

File: src/jdk.hotspot.agent/linux/native/libsaproc/ps_core.c
Comment (avu) [inline src/jdk.hotspot.agent/linux/native/libsaproc/ps_core.c:655]: Extra white spaces before 'if'

File: src/jdk.hotspot.agent/linux/native/libsaproc/ps_core.c
Comment (avu) [inline src/jdk.hotspot.agent/linux/native/libsaproc/ps_core.c:662]: Extra white-spaces before 'if'

File: src/jdk.hotspot.agent/linux/native/libsaproc/ps_core.c
Comment (avu) [inline src/jdk.hotspot.agent/linux/native/libsaproc/ps_core.c:717]: Extra white-spaces before 'if'

File: src/jdk.jpackage/share/native/common/tstrings.cpp
Comment (avu) [inline src/jdk.jpackage/share/native/common/tstrings.cpp:60]: Extra white-spaces before 'ret'

**Review by avu — CHANGES_REQUESTED** [review]

File: src/jdk.hotspot.agent/linux/native/libsaproc/ps_core.c
Comment (savoptik) [reply src/jdk.hotspot.agent/linux/native/libsaproc/ps_core.c]: done

File: src/jdk.hotspot.agent/linux/native/libsaproc/ps_core.c
Comment (savoptik) [reply src/jdk.hotspot.agent/linux/native/libsaproc/ps_core.c]: done

File: src/java.desktop/unix/native/libawt_xawt/awt/gtk2_interface.c
Comment (savoptik) [reply src/java.desktop/unix/native/libawt_xawt/awt/gtk2_interface.c]: done

File: src/jdk.hotspot.agent/linux/native/libsaproc/ps_core.c
Comment (savoptik) [reply src/jdk.hotspot.agent/linux/native/libsaproc/ps_core.c]: done

File: src/jdk.hotspot.agent/linux/native/libsaproc/ps_core.c
Comment (savoptik) [reply src/jdk.hotspot.agent/linux/native/libsaproc/ps_core.c]: done

File: src/jdk.jpackage/share/native/common/tstrings.cpp
Comment (savoptik) [reply src/jdk.jpackage/share/native/common/tstrings.cpp]: done

File: src/java.security.jgss/share/native/libj2gss/gssapi.h
Comment (savoptik) [reply src/java.security.jgss/share/native/libj2gss/gssapi.h]: done

Comment (savoptik) [reply src/java.security.jgss/share/native/libj2gss/gssapi.h]: This is rarely used in the code and is not the essence of the current changes.
If you introduce such changes, then throughout the code.
Moreover, this can lead to problems, such as, for example, here: https://bugs.openjdk.org/browse/JDK-8309225

**Comment by avu** [comment]
The latest changes look good to me.

**Review by avu — APPROVED** [review]

File: src/java.security.jgss/share/native/libj2gss/gssapi.h
Comment (savoptik) [reply src/java.security.jgss/share/native/libj2gss/gssapi.h]: I think you can create a separate ticket and pull request to discuss this issue.

**Comment by erikj79** [comment]
Build changes look ok.

File: make/modules/java.desktop/lib/Awt2dLibraries.gmk
Comment (djelinski) [inline make/modules/java.desktop/lib/Awt2dLibraries.gmk:235]: please group the disabled warnings by compiler (gcc, then clang, then clang_aix)

**Review by djelinski — COMMENTED** [review]
Please update copyright years where they aren't at 2023 yet.

I verified these changes on Ubuntu 20.04 + clang 10; they are both necessary and sufficient to make the build pass with warnings as errors. I saw one more warning:
```
clang: warning: argument unused during compilation: '-static-libstdc++' [-Wunused-command-line-argument]
```
in `BUILD_SYSLOOKUPLIB_link.log`, but that one does not break the build.

I'm going to approve this once you fix the mentioned issues. Thanks for working on this!

File: make/modules/java.desktop/lib/Awt2dLibraries.gmk
Comment (savoptik) [reply make/modules/java.desktop/lib/Awt2dLibraries.gmk:246]: Done

File: make/modules/java.desktop/lib/Awt2dLibraries.gmk
Comment (djelinski) [inline make/modules/java.desktop/lib/Awt2dLibraries.gmk:241]:
```suggestion
        DISABLED_WARNINGS_gcc_GLXSurfaceData.c := unused-function, \
        DISABLED_WARNINGS_gcc_gtk2_interface.c := parentheses type-limits, \
        DISABLED_WARNINGS_gcc_gtk3_interface.c := parentheses type-limits unused-function, \
        DISABLED_WARNINGS_gcc_OGLBufImgOps.c := format-nonliteral, \
        DISABLED_WARNINGS_gcc_OGLPaints.c := format-nonliteral, \
        DISABLED_WARNINGS_gcc_sun_awt_X11_GtkFileDialogPeer.c := parentheses, \
        DISABLED_WARNINGS_gcc_X11SurfaceData.c := implicit-fallthrough pointer-to-int-cast, \
```

File: make/modules/java.desktop/lib/Awt2dLibraries.gmk
Comment (djelinski) [inline make/modules/java.desktop/lib/Awt2dLibraries.gmk:260]: move this one closer to the other clang (non-aix) suppressions

Comment (savoptik) [reply make/modules/java.desktop/lib/Awt2dLibraries.gmk]: Done.

File: make/modules/java.desktop/lib/Awt2dLibraries.gmk
Comment (djelinski) [reply make/modules/java.desktop/lib/Awt2dLibraries.gmk:241]: Please remove the extra spaces.

Comment (savoptik) [reply make/modules/java.desktop/lib/Awt2dLibraries.gmk]: Done.

**Review by djelinski — APPROVED** [review]
LGTM

**Pull request closed without merging**
